plugins/services/src/js/utils/ServiceUtil.js

```diff
@@ -44,13 +47,30 @@ const getFindPropertiesRecursive = function(service, item) {
   }, {});
 };
 
+const getFrameworkImages = function(packageName) {
```

Comment:
I tried to put this in `FrameworkUtil`, but for whatever reason unit tests for `UniversePackagesList` fail when I import `CosmosPackagesStore` into `FrameworkUtil`.

Reply:
> TL;DR there is a better solution (IMHO) using the data layer that does not violate our architecture.

I am with Julian, `structs` should not have dependencies on how the data is fetched, and `utils` should be pure functions from the external perspective. That is an invariant that I believe helps us have a maintainable system, and I would like to keep it that way. 

I tried to provide a solution to having the cosmos dependency, by thinking what if I pass the result of `CosmosPackagesStore.getPackageImages()` as imagesDatabase to`getFrameworkImage(packageName, imageDatabase)`. But the `getFramework` method is called by `ServiceUtil.createServiceFromResponse`, which is then called by the ServiceTree `constructor`. There I stopped because I have no idea on how to tackle that simply. It seems to me that the solution is to make the DCOSStore aware of changes on package icons, it also looks to me that other places might build the tree. Also, the logic to build the tree is optimized via cache so we have to consider that the icons do not invalidate the cache unnecessarily..... That is altogether not simple.

This dependency of Cosmos on the util is for me a show stopper. I would gladly be overriden by the team if they think that is ok. I would also love if @nLight or @orlandohohmeier just jumped in with a simpler solution.

On the case, nothing simpler comes or the team thinks this dependency is fine (which I don't). I would gladly help you either way you chose to go: finish the direction you started or shift gears and still achieve that using the data layer.

*The data layer solution for me* is to create a mediator component who fetches the image just once using a ReplaySubject to cache the image data. We can use the CosmosStore to save time, but ideally, we should make the request ourselves using streams. The mediator component would be provided right before passing the service to the table, if it can find a cached image for the service, it passes it down, if It cannot it returns an empty image or placeholder. That part of the Service struct could require some thinking, but we at least avoid dealing with the tree.

Reply:
On an aside. To my own dismay, I looked into all the Utils and there is a single use of a Store in a util which I wrote when fixing the performance of the ServicePage. Is totally not needed as the callers of the `mergeTaskData`could pass DCOSStore.taskLookupTable.

Reply:
Thank you for your detailed feedback @Fabs!

I feel we should definitely go for the the better solution, instead of the ship-it that violates our architecture. Given that this ticket is a blocker for 1.12, I am not sure how much time it would take for me to implement the data layer solution. I wonder if we should just ship this solution and create a tech debt ticket to be addressed later.

@dcos/frontend-dev what does everyone think the course of action should be here?

Reply:
What does happen if we do not ship it on 1.12? Will our users start seeing broken links everywhere?

Reply:
Yes, the label is removed for 1.12. I was informed we have roughly 6 weeks until then. Given we have time, I think we should go for the best solution.

Reply:
After speaking with @juliangieseke, we can instead accessing the `CosmosPackagesStore` in `MarathomStore#processMarathonGroups`.

Reply:
It would be the only place not in a `util` or `struct` - not sure what issues we'll face changing the store and which places I havent seen tho… Ideas anyone?